Додано функціонал класів Phone, Birthday, Email

Olehsem23 opened the pull request:
1.  Для перевірки на правильність введення номеру телефону, використовуємо регулярний вираз, що українські номери обов'язково мають починатися з '+380' і ще 9 цифр. 
2. Для перевірки на правильність введення дати народження, необхідно щоб ми записували ці об'єкти як об'єкти типу datetime у форматі '%d.%m.%Y'. Це необхідно щоб одразу запобігти введенню 'дивних' даних на кшталт "2/2/42", '50.20.2000' чи навіть '29.02.2001'. Також варто уникати дат народження, які є у майбутньому. Це враховано у методі add_birthday класу Record.
3. Для перевірки на правильність введення email пропоную використати регулярний вираз, який був у нас у автоперевірці.
4. Додано відповідні методи у клас Record та команди, які потрібно вводити у терміналі, які записують дані для вибраного користувача. 
5. Додано рядкове представлення для виводу даних по записах (це лише моя пропозиція: варто обговорити, як краще це зробити).

Olehsem23 committed:
Додано функціонал класів Phone, Birthday, Email

Olehsem23 committed:
Додано функцію для відображення запис користувача.

bohdan-zhemelko commented on classes.py:
```diff
@@ -1,30 +1,123 @@
+# обов'язково мають починатися з '+380' і ще 9 цифр. 
 class Phone(Field):
-    ...
+    @property
```
Розгляньте такий варіант перевірки введення:

```
class Phone(Field):

    @Field.value.setter
    def value(self, value):
        if not re.match(r'^\+380\d{9}$', value):
            raise ValueError('Wrong format. Phone number should be in the format +380XXXXXXXXX.')
        self._Field__value = value
```

bohdan-zhemelko commented on classes.py:
```diff
@@ -1,30 +1,123 @@
+# Також варто уникати дат народження, які є у майбутньому. Це враховано у методі add_birthday класу Record.
 class Birthday(Field):
-    ...
+    @property
```
Розгляньте такий варіант перевірки введення:

```
class Birthday(Field):

    @Field.value.setter
    def value(self, value):
        try:
            self._Field__value = datetime.strptime(value, '%d.%m.%Y').date()
        except ValueError:
            raise ValueError('Wrong format. Enter birthday in format dd.mm.YYYY. Days in range(31), month in range(12)')
```

bohdan-zhemelko commented on classes.py:
```diff
@@ -1,30 +1,123 @@
+
+# Для перевірки на правильність введення email пропоную використати регулярний вираз, який був у нас у автоперевірці.
+class Email(Field):
+    @property
```
Розгляньте такий варіант перевірки введення:

```
class Email(Field):

    @Field.value.setter
    def value(self, value):
        if not re.match(r'[A-Za-z]{1}[\w.]+@[a-zA-Z0-9]+\.[a-zA-Z]{2,}\b', value):
            raise ValueError('Wrong format. Email should be in the format xxxxxx@xxx.xx')
        self._Field__value = value
```

bohdan-zhemelko commented on classes.py:
```diff
@@ -1,30 +1,123 @@
+        return f"phone: {phone} is already registered for user {self.name}"
+
+    # Функція додає дату народження користувача. Враховано, що дата народження не може бути у майбутньому.
+    def add_birthday(self, birthday: Birthday):
```
Переробіть цю функцію, оскільки при вказуванні майбутньої дати воно всерівно присвоює цю дату користувачу, а повинно не дозволяти.

Ось як я проводив дебагінг:

**record = Record('name', '+380123456789')
birthday = Birthday('1.2.2024')
print(record.add_birthday(birthday))
record.add_birthday(birthday)
print(record)**

І ось, що отримав на виході:

![image](https://github.com/tru-ten/Personal_assistant/assets/74614544/ecd44929-e609-45f2-8a03-2f4cd9ec1d81)

bohdan-zhemelko commented on functions.py:
```diff
@@ -48,14 +90,20 @@ def show_all(args):
 
+
 HANDLERS = {
     'add': add_user,
```
Назвіть тут ключ `add user`

bohdan-zhemelko commented on functions.py:
```diff
@@ -48,14 +90,20 @@ def show_all(args):
+    'add birthday': add_birthday_command,
+    'add email': add_email_command,
     'show all': show_all,
+    'show': show_user_command,
```
Назвіть тут ключ `show user`

bohdan-zhemelko commented on functions.py:
```diff
@@ -35,6 +38,45 @@ def add_user(args):
 
+
+@error_handler
+def add_phone_command(*args):  # Додаємо номер телефону для вибраного користувача.
```
Видаліть символ `*` при вказуванні аргументів, оскільки `*args` в даному випадку спримається як кортеж, а не як список через функцію `error_handler`

bohdan-zhemelko commented on functions.py:
```diff
@@ -35,6 +38,45 @@ def add_user(args):
+
+
+@error_handler
+def add_birthday_command(*args):  # додаємо дату народження для користувача
```
Видаліть символ `*` при вказуванні аргументів, оскільки `*args` в даному випадку спримається як кортеж, а не як список через функцію `error_handler`

bohdan-zhemelko commented on functions.py:
```diff
@@ -35,6 +38,45 @@ def add_user(args):
+
+
+@error_handler
+def add_email_command(*args):  # додаємо e-mail для користувача
```
Видаліть символ `*` при вказуванні аргументів, оскільки `*args` в даному випадку спримається як кортеж, а не як список через функцію `error_handler`

bohdan-zhemelko commented on functions.py:
```diff
@@ -35,6 +38,45 @@ def add_user(args):
+
+
+@error_handler
+def show_user_command(*args):  # Пошук телефона вибраного користувача.
```
Видаліть символ `*` при вказуванні аргументів, оскільки `*args` в даному випадку спримається як кортеж, а не як список через функцію `error_handler`

bohdan-zhemelko commented on functions.py:
```diff
@@ -35,6 +38,45 @@ def add_user(args):
+    rec: Record = contact_book.get(str(name))
+    if rec:
+        return rec.add_phone(phone)
+    rec = Record(name)
```
Цей рядок і наступний потрібно видалити, тому що при роботі цього методу вони не знадобляться, оскільки запис з ім'ям користувача ми вже задаємо за допомогою функції `add user`

bohdan-zhemelko commented on functions.py:
```diff
@@ -35,6 +38,45 @@ def add_user(args):
+    rec: Record = contact_book.get(str(name))
+    if rec:
+        return rec.add_birthday(birthday)
+    rec = Record(name)
```
Цей рядок і наступний потрібно видалити, тому що при роботі цього методу вони не знадобляться, оскільки запис з ім'ям користувача ми вже задаємо за допомогою функції `add user`

bohdan-zhemelko commented on functions.py:
```diff
@@ -35,6 +38,45 @@ def add_user(args):
+    rec: Record = contact_book.get(str(name))
+    if rec:
+        return rec.add_email(email)
+    rec = Record(name)
```
Цей рядок і наступний потрібно видалити, тому що при роботі цього методу вони не знадобляться, оскільки запис з ім'ям користувача ми вже задаємо за допомогою функції `add user`

Olehsem23 committed:
Зроблено зміни у назві файлів та у функціоналах

Olehsem23 commented on personal-assistant.py:
```diff
@@ -1,4 +1,5 @@
-from functions import parse_input
+# Внаслідок перейменування пакетів оновлено імпорт пакетів
+from contact_book_functions import parse_input
 
```
Пропоную щоб ввід був трохи інформативнішим. і додав \n щоб була відстань між командами. так для сприйняття краще.

Olehsem23 commented on contact_book_functions.py:
```diff
@@ -0,0 +1,111 @@
+        except KeyError:
+            return "No user"
+        # під час виконання різних методів виникають різні помилки ValueError. Тому пропоную їх перехоплювати у методах, 
+        # щоб користувач знав у чому проблема. Крім того нам ще треба продумавти логіку для перехоплення помилок типу AttribiteError.
```
# Варто ще обробити помилки TypeError та AttribiteError. Вони точно будуть виникати під час роботи.

Olehsem23 commented on contact_book_functions.py:
```diff
@@ -0,0 +1,111 @@
+            return "No user"
+        # під час виконання різних методів виникають різні помилки ValueError. Тому пропоную їх перехоплювати у методах, 
+        # щоб користувач знав у чому проблема. Крім того нам ще треба продумавти логіку для перехоплення помилок типу AttribiteError.
+        except ValueError as e:
```
Команда, яка видає список всіх доступних команд, щоб користувач міг легше зорієнтуватися. Ну і навішав всім декораторів, щоб обробляли помилки.

Olehsem23 commented on contact_book_functions.py:
```diff
@@ -0,0 +1,111 @@
+    return "unknown_command"
+
+
+def exit(args):
```
Rename to 'exit_command'

Olehsem23 commented on contact_book_functions.py:
```diff
@@ -0,0 +1,111 @@
+
+
+HANDLERS = {
+    'add user': add_user,
```
Перероблено, щоб функції були ключами, а слова, якими ці команди викликаються - це tuple команд. Тепер можна дуже багато команд на виклик однієї команди навішати.

Olehsem23 commented on contact_book_functions.py:
```diff
@@ -0,0 +1,111 @@
+    return contact_book[args[0]]
+
+
+def show_all(args):
```
rename to 'show_all_command'

Olehsem23 commented on contact_book_functions.py:
```diff
@@ -0,0 +1,111 @@
+def add_user(args):
+    name = Name(args[0])
+    rec = Record(name)
+    if len(args)==2:
```
Додано фунції, які працюють з днями народженнями користувачів, та повертають відповідну інформацію.

Olehsem23 commented on contact_book_functions.py:
```diff
@@ -0,0 +1,111 @@
+
+HANDLERS = {
+    'add user': add_user,
+    'add phone': add_phone_command,
```
Додано звернення до функцій, які працюють з ДН.

Olehsem23 commented on contact_book_functions.py:
```diff
@@ -0,0 +1,111 @@
+        command = command.lstrip()
+        handler = HANDLERS[command.lower()]
+    except KeyError:
+        if args:
```
Перероблено парсер, на вигляд який пропонував ментор на лекції. На мій погляд, так зручніше працювати.

Olehsem23 commented on contact_book_classes.py:
```diff
@@ -0,0 +1,114 @@
+    def value(self, value):
+        try:
+            self._Field__value = datetime.strptime(value, '%d.%m.%Y').date()
+        except ValueError:
```
Додано перевірку на те, що дата народження не може бути у майбутньому.

Olehsem23 commented on contact_book_classes.py:
```diff
@@ -0,0 +1,114 @@
+            self.phones.append(phone)
+            return f"phone {phone} was added to contact {self.name}"
+        return f"phone: {phone} is already registered for user {self.name}"
+
```
Оновлено функцію, яка додає день народження користувача. умова про день народження у майбутньому перенесена в опис класу Birthday.

Olehsem23 commented on contact_book_classes.py:
```diff
@@ -0,0 +1,114 @@
+        self._Field__value = value
+
+
+class Record:
```
Додано функцію, яка повертає інформацію про кількість днів до дня народження користувача. Ще мені знадобилася додаткова фунція days_to_birthday_int_numbers, яка повертає лише числа. Вона нижче використовується, для методів класу AddressBook.

Olehsem23 commented on contact_book_classes.py:
```diff
@@ -0,0 +1,114 @@
+# Також варто уникати дат народження, які є у майбутньому. Це враховано у методі add_birthday класу Record.
+class Birthday(Field):
+
+    @Field.value.setter
```
Реалізовано функції, які також повертають списки працівників, котрі святкують ДН у вибраний період часу. Додатково фунції, які повертають список для поточного чи наступного тижня, та поточного та наступного місяця.

Olehsem23 commented on contact_book_functions.py:
```diff
@@ -0,0 +1,111 @@
+
+
+@error_handler
+def add_phone_command(args):  # Додаємо номер телефону для вибраного користувача.
```
все-таки потрібні ставити *args, бо без цього бере тільки перший символ і все.

Olehsem23 committed:
Пофіксив баги стосовно ДН і додав методи Birthday

bohdan-zhemelko commented on __pycache__/classes.cpython-311.pyc:
Як я розумію цей файл виник коли ви робили дебагінг, тому можете його видалити

bohdan-zhemelko commented on __pycache__/contact_book_classes.cpython-311.pyc:
Як я розумію цей файл виник коли ви робили дебагінг, тому можете його видалити

bohdan-zhemelko commented on __pycache__/contact_book_functions.cpython-311.pyc:
Як я розумію цей файл виник коли ви робили дебагінг, тому можете його видалити

bohdan-zhemelko commented on __pycache__/functions.cpython-311.pyc:
Як я розумію цей файл виник коли ви робили дебагінг, тому можете його видалити

bohdan-zhemelko commented on personal-assistant.py:
```diff
@@ -1,14 +1,18 @@
-        if not result:
-            print('Exit')
-            break
+        user_input = input('\nEnter your command and args: ')
```
Варто додати до тексту input, що є така команда як help при натисканні якої видасть список команд. А то так виходить, що запускається бот, який вимагає команду, а користувач не знає які команди існують

Olehsem23 replied:
Прийнято. Хоча я думаю, що краще це додати у привітальному повідомленні. Коли бот розказує про себе, що я вмію робити, якщо потрібно взнати перелік команд натисніть help. Хоча можна і там, і там це додати. Я напишу

Olehsem23 committed:
fixed some small bugs

Пофіксив стосовно ДН, коли нема жодного ДН в списку викидало з програми (обробив помилку).
Додав привітальний текст, щоб користувач мав змогу ознайомитися зі списком доступних операцій.

KossKokos committed:
add address

bohdan-zhemelko commented on __pycache__/contact_book_classes.cpython-311.pyc:
Слід видалити

bohdan-zhemelko commented on __pycache__/contact_book_functions.cpython-311.pyc:
Слід видалити

bohdan-zhemelko commented on contact_book_functions.py:
```diff
@@ -0,0 +1,159 @@
+        except ValueError as e:
+            return e
+        except IndexError:
+            return 'Not enough parameters'
```
Тут слід написати так `return 'First you should enter the username and, if necessary, the required parameter'`

bohdan-zhemelko commented on contact_book_functions.py:
```diff
@@ -0,0 +1,159 @@
+            result = func(*args)
+            return result
+        except KeyError:
+            return f"No user with name {args[0]}. You need firstly create "
```
Тут слід написати так `return 'No user'`

Olehsem23 commented on contact_book_functions.py:
```diff
@@ -0,0 +1,159 @@
+        # під час виконання різних методів виникають різні помилки ValueError. Тому пропоную їх перехоплювати у методах, 
+        # щоб користувач знав у чому проблема. Крім того нам ще треба продумавти логіку для перехоплення помилок типу AttribiteError.
+        except ValueError as e:
+            return e
```
Додано функцію start

Olehsem23 commented on contact_book_functions.py:
```diff
@@ -0,0 +1,159 @@
+
+
+@error_handler
+def exit_command():
```
При виході з програми обовязково всі зміни зберігаються до файлу contact_book.bin

KossKokos committed:
bugfix/method-add_address

KossKokos committed:
bugfix/method-add_address

Olehsem23 commented on contact_book_functions.py:
```diff
@@ -0,0 +1,159 @@
+        print(f'There are {len(contact_book)} users in address book')
+        return contact_book
+
+
```
Додав ще й до кожної команди номер. Тепер користувач коли викличе 'help', то він побачить що можна програму викликати різними методами. Наприклад, щоб додати телефон користувачу, він може набрати 'add phone', а може просто набрати '12' і це викличе програму на додавання телефону. Так само з усіма іншими.

Olehsem23 commented on contact_book_classes.py:
```diff
@@ -0,0 +1,221 @@
+            return f"There are {diff_days} days left until the {self.name}'s {diff_years} birthday"
+        diff_days = (bd_next_year - today).days
+        return f"There are {diff_days} days left until the {self.name}'s {diff_years + 1} birthday"
+    
```
Записує все у файл

Olehsem23 commented on contact_book_classes.py:
```diff
@@ -0,0 +1,221 @@
+        if not self.birthday:
+            return -1 # якщо дати народження нема, то повертає -1
+        today = datetime.now().date()
+        bd_current_year = self.birthday.value.replace(year=today.year)
```
вивантажує всю інформацію за файлу для роботи.

Olehsem23 commented on contact_book_classes.py:
```diff
@@ -0,0 +1,221 @@
+        diff_days = (bd_next_year - today).days
+        return diff_days
+        
+    # Рядкове представлення для одного запису у contact_book 
```
Метод шукає задану послідовність символів по всій книзі. Виводить список де знайшло.

Olehsem23 committed:
Додано пошук та серіалізація.

Додано метод на пошук збігів у цілій контактній книзі. Додано серіалізацію та десеріалізацію. Одразу при запуску бота вивантажується інформація з бінарного файлу, і при закінченні роботи тобто при введенні програми "exit" всі зміни завантажуються назад у бінарний файл.
Також додано короткі команди для виклику команд. З ними можна ознайомитися при виклику команди 'help'. Тепер користувач може вводити як коротку команду з двох цифр, так і повний текст команди.

KossKokos committed:
bugfix/Country-setter

Olehsem23 committed:
Оброблено KeyError та IndexError.

Olehsem23 committed:
Виправлено KeyError та IndexError

bohdan-zhemelko committed:
Merge branch 'faeture/classes-address+func-add_address' into feature/class-Phone-Email-Birthday
